### PR TITLE
loader: Move Loader to new pkg loader

### PR DESCRIPTION
**Pull request by tuxcanfly**
As mentioned in #463 we need a way to start/stop the ticketbuyer purchaser over gRPC.

This PR allows the loader to be re-usable from `pkg loader`  so as to handle ticketbuyer services to be stopped/started similar to the current handling of wallet load/unload.

### CODE REVIEW

**Comment by tuxcanfly** [comment]
Rebased and improved `ticketbuyer-rpc` onto this branch. This branch now implements a `ticketbuyer` service using the new `pkg loader`. Ticket purchasing can be enabled on the fly using the `StartTicketPurchase` rpc call which requires the passphrase and the config. All config params can be tweaked with a corresponding rpc call.

**Comment by tuxcanfly** [comment]
Rebased, fixed conflicts. Running tests to make sure nothing was lost during rebase.

**Comment by jrick** [comment]
Sorry for not being able to review this in time for the release.  We'll do so and get it in soon after release is pushed.

**Comment by tuxcanfly** [comment]
OK, thanks. Will fix a couple of issues found during testing and polish it up for review.

File: dcrwallet.go [hunk 1]
```diff
@@ -91,7 +90,7 @@ func walletMain() error {
 	}
 
 	dbDir := networkDir(cfg.AppDataDir, activeNet.Params)
-	stakeOptions := &wallet.StakeOptions{
+	stakeOptions := &ldr.StakeOptions{
```
Comment (jrick) [inline dcrwallet.go]: why was this moved to the loader package?

File: dcrwallet.go [hunk 2]
```diff
@@ -100,14 +99,14 @@ func walletMain() error {
 		PruneTickets:            cfg.PruneTickets,
 		AddressReuse:            cfg.ReuseAddresses,
 		TicketAddress:           cfg.TicketAddress,
-		TicketMaxPrice:          cfg.tbCfg.MaxPriceAbsolute,
+		TicketMaxPrice:          cfg.tbCfg.MaxPriceAbsolute.ToCoin(),
```
Comment (jrick) [inline dcrwallet.go]: definitely don't want this. stick to Amount for anything internal

File: loader/disksync.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) 2015 The btcsuite developers
```
Comment (jrick) [inline loader/disksync.go]: this file should be renamed to something more meaningful, or move the function to another file.

File: loader/loader.go [hunk 4]
```diff
@@ -93,59 +85,42 @@ func NewLoader(chainParams *chaincfg.Params, dbDirPath string,
 	}
 }
 
-// onLoaded executes each added callback and prevents loader from loading any
+// onLoaded executes each added callback and prevents from loading any
```
Comment (jrick) [inline loader/loader.go]: > prevents the loader

File: loader/loader.go [hunk 5]
```diff
@@ -93,59 +85,42 @@ func NewLoader(chainParams *chaincfg.Params, dbDirPath string,
...
-func (l *Loader) onUnload(w *Wallet) {
-	for _, fn := range l.unloadCallbacks {
-		fn(w)
-	}
-	l.unloadCallbacks = nil // not needed anymore
+	l.callbacks = nil // not needed anymore
 }
 
-// RunAfterLoad adds a function to be executed when the loader creates or opens
+// RunAfterLoad adds a function to be executed when the creates or opens
```
Comment (jrick) [inline loader/loader.go]: nope, this was right before.

File: loader/loader.go [hunk 6]
```diff
@@ -265,7 +240,7 @@ func (l *Loader) OpenExistingWallet(pubPassphrase []byte, canConsolePrompt bool)
 	return w, nil
 }
 
-// WalletExists returns whether a file exists at the loader's database path.
+// WalletExists returns whether a file exists at the s database path.
```
Comment (jrick) [inline loader/loader.go]: uh, no.

File: loader/loader.go [hunk 7]
```diff
@@ -307,6 +281,69 @@ func (l *Loader) UnloadWallet() error {
...
+		return ErrWalletNotLoaded
+	}
+	w := l.wallet
+	p, err := ticketbuyer.NewTicketPurchaser(ticketbuyerCfg, l.chainClient, w, l.chainParams)
+	if err != nil {
+		return err
+	}
+	if passphrase != nil {
+		var unlockAfter <-chan time.Time
+		err = w.Unlock(passphrase, unlockAfter)
```
Comment (jrick) [inline loader/loader.go]: can just pass nil in here instead of declaring a channel variable for it.

File: loader/loader.go [hunk 8]
```diff
@@ -307,6 +281,69 @@ func (l *Loader) UnloadWallet() error {
...
+
+	if l.purchaseManager != nil {
+		l.ntfnClient.Done()
+		l.purchaseManager.Stop()
+		l.purchaseManager.WaitForShutdown()
+		l.purchaseManager = nil
+		return nil
+	}
+
+	return ErrTicketBuyerStopped
```
Comment (jrick) [inline loader/loader.go]: please reverse these checks, error out early if the purchase manager is nil, and put the actual stopping code at this indent level below it.

File: loader/loader.go [hunk 9]
```diff
@@ -307,6 +281,69 @@ func (l *Loader) UnloadWallet() error {
...
+}
+
+// PurchaseManager returns the ticket purchaser instance.
+func (l *Loader) PurchaseManager() *ticketbuyer.PurchaseManager {
+	defer l.mu.Unlock()
```
Comment (jrick) [inline loader/loader.go]: silly use of defer here

File: loader/loader.go [hunk 10]
```diff
@@ -307,6 +281,69 @@ func (l *Loader) UnloadWallet() error {
...
+		l.purchaseManager.Stop()
+		l.purchaseManager.WaitForShutdown()
+		l.purchaseManager = nil
+		return nil
+	}
+
+	return ErrTicketBuyerStopped
+}
+
+// PurchaseManager returns the ticket purchaser instance.
```
Comment (jrick) [inline loader/loader.go]: it *might* return the ticket purchaser instance.  or it might not if it wasn't set.  comment should at least call out that this returns an optional and the caller must check for nil.

File: rpc/api.proto [hunk 11]
```diff
@@ -459,3 +476,93 @@ message DecodeSeedRequest {
 message DecodeSeedResponse {
 	bytes decoded_seed = 1;
 }
+
+message StartTicketPurchaseRequest {
+	bytes passphrase = 1;
+	string account_name           = 2;
```
Comment (jrick) [inline rpc/api.proto]: none of the other field positions are aligned, don't do it here.

they're not even aligned correctly.

File: rpc/api.proto [hunk 12]
```diff
@@ -459,3 +476,93 @@ message DecodeSeedRequest {
...
+message StartTicketPurchaseResponse {}
+
+message StopTicketPurchaseRequest {}
+message StopTicketPurchaseResponse {}
+
+message TicketBuyerConfigRequest{
+}
+
+message TicketBuyerConfigResponse{
+    string account_name           = 1;
```
Comment (jrick) [inline rpc/api.proto]: no alignment

File: rpc/api.proto [hunk 13]
```diff
@@ -64,6 +64,23 @@ service SeedService {
 	rpc DecodeSeed (DecodeSeedRequest) returns (DecodeSeedResponse);
 }
 
+service TicketBuyerService {
+    rpc StartTicketPurchase (StartTicketPurchaseRequest) returns (StartTicketPurchaseResponse);
+    rpc StopTicketPurchase (StopTicketPurchaseRequest) returns (StopTicketPurchaseResponse);
+
+    rpc Config (TicketBuyerConfigRequest) returns (TicketBuyerConfigResponse);
```
Comment (jrick) [inline rpc/api.proto]: the request and response message type names should be the name of the method followed by Request or Response, like all of the other ones.  I would change the method name in this case (to TicketBuyerConfig).

File: rpc/documentation/api.md [hunk 14]
```diff
@@ -1476,3 +1478,27 @@ The user input can be either a hexadecimal string or a mnemonic word list.
 - `InvalidArgument`: The input is invalid.
 
 **Stability:** Unstable
+
+#### `StartTicketPurchase`
```
Comment (jrick) [inline rpc/documentation/api.md]: Missing tons of doco here.

File: rpc/api.proto [hunk 15]
```diff
@@ -64,6 +64,23 @@ service SeedService {
 	rpc DecodeSeed (DecodeSeedRequest) returns (DecodeSeedResponse);
 }
 
+service TicketBuyerService {
+    rpc StartTicketPurchase (StartTicketPurchaseRequest) returns (StartTicketPurchaseResponse);
+    rpc StopTicketPurchase (StopTicketPurchaseRequest) returns (StopTicketPurchaseResponse);
```
Comment (jrick) [inline rpc/api.proto]: Please define these as part of the WalletLoaderService since this requires touching the loader and to keep it consistent with starting the wallet.

File: rpc/rpcserver/server.go [hunk 16]
```diff
@@ -139,6 +141,201 @@ func decodeAddress(a string, params *chaincfg.Params) (dcrutil.Address, error) {
 type versionServer struct {
 }
 
+// ticketBuyerServer provides ticket buyer services for RPC clients.
+type ticketBuyerServer struct {
+	loader         *loader.Loader
```
Comment (jrick) [inline rpc/rpcserver/server.go]: as commented above, loading/unloading should be done by the WalletLoaderService

File: rpc/rpcserver/server.go [hunk 17]
```diff
@@ -139,6 +141,201 @@ func decodeAddress(a string, params *chaincfg.Params) (dcrutil.Address, error) {
...
+	ticketbuyerCfg *ticketbuyer.Config
+}
+
+// StartTicketPurchase enables ticket purchases in the ticket buyer.
```
Comment (jrick) [inline rpc/rpcserver/server.go]: I actually don't require comments on these methods since they are invisible to godoc users.  all documentation should be placed in the api specification markdown document.

File: dcrwallet.go (hunk 1 above)
Comment (tuxcanfly) [inline dcrwallet.go]: AFAIK, it's only required in loader pkg, so I had moved it here.

File: dcrwallet.go (hunk 2 above)
Comment (tuxcanfly) [inline dcrwallet.go]: OK.

File: loader/disksync.go (hunk 3 above)
Comment (tuxcanfly) [inline loader/disksync.go]: Will do.

File: rpc/rpcserver/server.go [hunk 18]
```diff
@@ -139,6 +141,201 @@ func decodeAddress(a string, params *chaincfg.Params) (dcrutil.Address, error) {
...
+	}, nil
+}
+
+// SetAccountName sets the account name to use for purchasing tickets.
+func (t *ticketBuyerServer) SetAccountName(ctx context.Context, req *pb.SetAccountNameRequest) (
+	*pb.SetAccountNameResponse, error) {
+
+	pm := t.loader.PurchaseManager()
+	if pm == nil {
+		return nil, grpc.Errorf(codes.FailedPrecondition, "ticket buyer is not running")
```
Comment (jrick) [inline rpc/rpcserver/server.go]: since this pattern is repeated so much please make a helper function to return a required purchase manager or error with a grpc error code

File: loader/loader.go (hunk 4 above)
Comment (tuxcanfly) [inline loader/loader.go]: Sorry, search and replace gone wrong.

File: loader/loader.go (hunk 8 above)
Comment (tuxcanfly) [inline loader/loader.go]: OK.

File: loader/loader.go (hunk 9 above)
Comment (tuxcanfly) [inline loader/loader.go]: OK, I thought it would prevent a allocation.

File: loader/loader.go (hunk 10 above)
Comment (tuxcanfly) [inline loader/loader.go]: Thanks, will update.

File: rpc/api.proto (hunk 12 above)
Comment (tuxcanfly) [inline rpc/api.proto]: Ok, will do.

File: rpc/documentation/api.md (hunk 14 above)
Comment (tuxcanfly) [inline rpc/documentation/api.md]: Sorry, I was thinking of getting to this once all the RPC commands are in. Will update that.

File: rpc/rpcserver/server.go (hunk 16 above)
Comment (tuxcanfly) [inline rpc/rpcserver/server.go]: OK.

File: dcrwallet.go [hunk 19]
```diff
@@ -144,6 +139,7 @@ func walletMain() error {
 			startPromptPass(w)
 		}
 		startWalletRPCServices(w, rpcs, legacyRPCServer)
+		startTicketBuyerService(rpcs, loader, &cfg.tbCfg)
```
Comment (jrick) [inline dcrwallet.go]: this function can be removed and it should be performed in startWalletRPCServices.

File: ticketbuyer/buyer.go [hunk 20]
```diff
@@ -104,6 +101,198 @@ type TicketPurchaser struct {
...
+// TicketAddress returns the address to send ticket outputs to.
+func (t *TicketPurchaser) TicketAddress() dcrutil.Address {
+	t.purchaserMtx.Lock()
+	ticketAddress := t.ticketAddress
+	t.purchaserMtx.Unlock()
+	return ticketAddress
+}
+
+// SetTicketAddress sets the address to send ticket outputs to.
+func (t *TicketPurchaser) SetTicketAddress(ticketAddress string) error {
```
Comment (jrick) [inline ticketbuyer/buyer.go]: take the decoded address not a string.  the rpc server should do all conversions and return the correct errors for invalid parameters.

File: ticketbuyer/buyer.go [hunk 21]
```diff
@@ -104,6 +101,198 @@ type TicketPurchaser struct {
...
+// AccountName returns the account name to use for purchasing tickets.
+func (t *TicketPurchaser) AccountName() string {
+	t.purchaserMtx.Lock()
+	accountName := t.accountName
+	t.purchaserMtx.Unlock()
+	return accountName
+}
+
+// SetAccountName sets the account name to use for purchasing tickets.
+func (t *TicketPurchaser) SetAccountName(accountName string) {
```
Comment (jrick) [inline ticketbuyer/buyer.go]: this needs some sort of validation that the account exists.

We should also be using account numbers to refer to accounts, never names.  Do the conversion from names to numbers in the rpc server handlers.

Comment (jrick) [inline ticketbuyer/buyer.go]: Oh and the validation issue becomes a non-issue if we do that conversion in the rpc server

File: ticketbuyer/buyer.go [hunk 22]
```diff
@@ -104,6 +101,198 @@ type TicketPurchaser struct {
...
+// PoolAddress returns the pool address where ticket fees are sent.
+func (t *TicketPurchaser) PoolAddress() dcrutil.Address {
+	t.purchaserMtx.Lock()
+	poolAddress := t.poolAddress
+	t.purchaserMtx.Unlock()
+	return poolAddress
+}
+
+// SetPoolAddress sets the pool address where ticket fees are sent.
+func (t *TicketPurchaser) SetPoolAddress(poolAddress string) error {
```
Comment (jrick) [inline ticketbuyer/buyer.go]: don't use string

File: ticketbuyer/buyer.go [hunk 23]
```diff
@@ -308,14 +486,11 @@ func (t *TicketPurchaser) Purchase(height int64) (*PurchaseStats, error) {
 	// Set the max price to the configuration parameter that is lower
 	// Absolute or relative max price
 	var maxPriceAmt dcrutil.Amount
-	if t.cfg.MaxPriceAbsolute > 0 && t.cfg.MaxPriceAbsolute < avgPrice*t.cfg.MaxPriceRelative {
-		maxPriceAmt, err = dcrutil.NewAmount(t.cfg.MaxPriceAbsolute)
-		if err != nil {
-			return ps, err
-		}
+	if t.MaxPriceAbsolute().ToCoin() > 0 && t.MaxPriceAbsolute().ToCoin() < avgPrice*t.MaxPriceRelative() {
+		maxPriceAmt = t.MaxPriceAbsolute()
```
Comment (jrick) [inline ticketbuyer/buyer.go]: this is incorrect, read the value once and do all checking on it before using it.

File: ticketbuyer/buyer.go [hunk 24]
```diff
@@ -399,14 +574,14 @@ func (t *TicketPurchaser) Purchase(height int64) (*PurchaseStats, error) {
 	// for by the user.
 	feeToUse := chainFee * t.cfg.FeeTargetScaling
 	log.Tracef("Average ticket fee: %.8f DCR", chainFee)
-	if feeToUse > t.cfg.MaxFee {
+	if feeToUse > t.MaxFee().ToCoin() {
 		log.Infof("Not buying because max fee exceed: (max fee: %.8f DCR,  scaled fee: %.8f DCR)",
-			t.cfg.MaxFee, feeToUse)
+			t.MaxFee().ToCoin(), feeToUse)
```
Comment (jrick) [inline ticketbuyer/buyer.go]: read once

File: ticketbuyer/buyer.go [hunk 25]
```diff
@@ -334,15 +509,15 @@ func (t *TicketPurchaser) Purchase(height int64) (*PurchaseStats, error) {
...
+	account, err := t.wallet.AccountNumber(t.AccountName())
 	if err != nil {
 		return ps, err
 	}
 	bal, err := t.wallet.CalculateAccountBalance(account, 0)
 	if err != nil {
 		return ps, err
 	}
-	log.Debugf("Spendable balance for account '%s': %v", t.cfg.AccountName, bal.Spendable)
+	log.Debugf("Spendable balance for account '%s': %v", t.AccountName(), bal.Spendable)
```
Comment (jrick) [inline ticketbuyer/buyer.go]: read once

File: ticketbuyer/buyer.go [hunk 26]
```diff
@@ -399,14 +574,14 @@ func (t *TicketPurchaser) Purchase(height int64) (*PurchaseStats, error) {
...
 		log.Infof("Not buying because max fee exceed: (max fee: %.8f DCR,  scaled fee: %.8f DCR)",
-			t.cfg.MaxFee, feeToUse)
+			t.MaxFee().ToCoin(), feeToUse)
 		return ps, nil
 	}
-	if feeToUse < t.cfg.MinFee {
+	if feeToUse < t.cfg.MinFee.ToCoin() {
 		log.Debugf("Using min ticket fee: %.8f DCR (scaled fee: %.8f DCR)", t.cfg.MinFee, feeToUse)
-		feeToUse = t.cfg.MinFee
+		feeToUse = t.cfg.MinFee.ToCoin()
```
Comment (jrick) [inline ticketbuyer/buyer.go]: read once

Comment (jrick) [inline ticketbuyer/buyer.go]: also looks concurrent unsafe

File: ticketbuyer/buyer.go [hunk 27]
```diff
@@ -421,13 +596,10 @@ func (t *TicketPurchaser) Purchase(height int64) (*PurchaseStats, error) {
 	// Set the balancetomaintain to the configuration parameter that is higher
 	// Absolute or relative balance to maintain
 	var balanceToMaintainAmt dcrutil.Amount
-	if t.cfg.BalanceToMaintainAbsolute > 0 && t.cfg.BalanceToMaintainAbsolute >
+	if t.cfg.BalanceToMaintainAbsolute > 0 && t.cfg.BalanceToMaintainAbsolute.ToCoin() >
```
Comment (jrick) [inline ticketbuyer/buyer.go]: kill ToCoin with fire

File: ticketbuyer/buyer.go [hunk 28]
```diff
@@ -421,13 +596,10 @@ func (t *TicketPurchaser) Purchase(height int64) (*PurchaseStats, error) {
...
 		bal.Total.ToCoin()*t.cfg.BalanceToMaintainRelative {
 
-		balanceToMaintainAmt, err = dcrutil.NewAmount(t.cfg.BalanceToMaintainAbsolute)
-		if err != nil {
-			return ps, err
-		}
+		balanceToMaintainAmt = t.cfg.BalanceToMaintainAbsolute
```
Comment (jrick) [inline ticketbuyer/buyer.go]: looks suspiciously concurrent unsafe

File: ticketbuyer/buyer.go [hunk 29]
```diff
@@ -577,8 +749,8 @@ func (t *TicketPurchaser) Purchase(height int64) (*PurchaseStats, error) {
 	// If an address wasn't passed, create an internal address in
 	// the wallet for the ticket address.
 	var ticketAddress dcrutil.Address
-	if t.ticketAddress != nil {
-		ticketAddress = t.ticketAddress
+	if t.TicketAddress() != nil {
+		ticketAddress = t.TicketAddress()
```
Comment (jrick) [inline ticketbuyer/buyer.go]: read once

File: ticketbuyer/buyer.go [hunk 30]
```diff
@@ -623,7 +795,7 @@ func (t *TicketPurchaser) Purchase(height int64) (*PurchaseStats, error) {
 	if err != nil {
 		return ps, err
 	}
-	log.Debugf("Usable balance for account '%s' after purchases: %v", t.cfg.AccountName, bal.Spendable)
+	log.Debugf("Usable balance for account '%s' after purchases: %v", t.AccountName(), bal.Spendable)
```
Comment (jrick) [inline ticketbuyer/buyer.go]: account was read before, need to use the same one here

**Comment by tuxcanfly** [comment]
Squashed. Limiting this PR to loader refactor specific changes.

File: config.go [hunk 31]
```diff
@@ -1014,7 +1014,6 @@ func loadConfig() (*config, []string, error) {
 		MaxInMempool:              cfg.TBOpts.MaxInMempool,
 		PoolAddress:               cfg.PoolAddress,
 		PoolFees:                  cfg.PoolFees,
-		PriceTarget:               cfg.TBOpts.PriceTarget.ToCoin(),
```
Comment (alexlyp) [inline config.go]: This has been deprecated, but not yet removed.

File: loader/loader.go [hunk 32]
```diff
@@ -307,6 +292,70 @@ func (l *Loader) UnloadWallet() error {
...
+	// Already running?
+	if l.purchaseManager != nil {
+		return ErrTicketBuyerStarted
+	}
+
+	if l.wallet == nil {
+		return ErrWalletNotLoaded
+	}
+	w := l.wallet
+	p, err := ticketbuyer.NewTicketPurchaser(ticketbuyerCfg, l.chainClient, w, l.chainParams)
```
Comment (jrick) [inline loader/loader.go]: what happens if the chain client was not previously set?

File: rpc/api.proto [hunk 33]
```diff
@@ -57,6 +57,8 @@ service WalletLoaderService {
 	rpc DiscoverAddresses (DiscoverAddressesRequest) returns (DiscoverAddressesResponse);
 	rpc SubscribeToBlockNotifications (SubscribeToBlockNotificationsRequest) returns (SubscribeToBlockNotificationsResponse);
 	rpc FetchHeaders(FetchHeadersRequest) returns (FetchHeadersResponse);
+    rpc StartTicketPurchase (StartTicketPurchaseRequest) returns (StartTicketPurchaseResponse);
```
Comment (jrick) [inline rpc/api.proto]: these changes should be split to another commit

File: rpc/documentation/api.md [hunk 34]
```diff
@@ -377,6 +377,8 @@ The service provides the following methods:
 - [`TransactionNotifications`](#transactionnotifications)
 - [`SpentnessNotifications`](#spentnessnotifications)
 - [`AccountNotifications`](#accountnotifications)
+- [`StartTicketPurchase`](#startticketpurchase)
```
Comment (jrick) [inline rpc/documentation/api.md]: same here, split them out into a future commit

File: dcrwallet.go [hunk 35]
```diff
@@ -309,7 +305,14 @@ func rpcClientConnectLoop(legacyRPCServer *legacyrpc.Server, loader *wallet.Load
 			if legacyRPCServer != nil {
 				legacyRPCServer.SetChainServer(chainClient)
 			}
-			c <- chainClient
+			if cfg.EnableTicketBuyer {
+				loader.SetChainClient(chainClient.Client)
+				err := loader.StartTicketPurchase(nil, &cfg.tbCfg)
```
Comment (jrick) [inline dcrwallet.go]: this seems to be the only call to StartTicketPurchase so I would move the chain client into its parameter and then we don't need to deal with the possibility of forgetting to set it after a refactor later.

File: dcrwallet.go [hunk 36]
```diff
@@ -309,7 +305,14 @@ func rpcClientConnectLoop(legacyRPCServer *legacyrpc.Server, loader *wallet.Load
...
+				if err != nil {
+					log.Errorf("Unable to start ticket buyer: %v", err)
+					return
```
Comment (jrick) [inline dcrwallet.go]: returning here seems questionable

File: dcrwallet.go [hunk 37]
```diff
@@ -322,6 +325,7 @@ func rpcClientConnectLoop(legacyRPCServer *legacyrpc.Server, loader *wallet.Load
 		})
 
 		chainClient.WaitForShutdown()
+		loader.StopTicketPurchase()
```
Comment (jrick) [inline dcrwallet.go]: please assign the error to _ so it's explicit that it is being ignored (comment would be nice too)

File: dcrwallet.go [hunk 38]
```diff
@@ -181,8 +176,9 @@ func walletMain() error {
 	// before exiting.  Interrupt handlers run in LIFO order, so the wallet
 	// (which should be closed last) is added first.
 	addInterruptHandler(func() {
+		loader.StopTicketPurchase()
```
Comment (jrick) [inline dcrwallet.go]: please assign the error to _ so it's explicit that it is being ignored (comment would be nice too)

File: config.go (hunk 31 above)
Comment (tuxcanfly) [inline config.go]: Thanks, restored.

File: rpc/api.proto (hunk 33 above)
Comment (tuxcanfly) [inline rpc/api.proto]: OK, made another commit for the RPCs.

File: rpc/documentation/api.md (hunk 34 above)
Comment (tuxcanfly) [inline rpc/documentation/api.md]: Done.

File: dcrwallet.go (hunk 38 above)
Comment (tuxcanfly) [inline dcrwallet.go]: OK, updated.

File: dcrwallet.go (hunk 36 above)
Comment (tuxcanfly) [inline dcrwallet.go]: Removed.

File: dcrwallet.go (hunk 37 above)
Comment (tuxcanfly) [inline dcrwallet.go]: Done.

File: dcrwallet.go (hunk 35 above)
Comment (tuxcanfly) [inline dcrwallet.go]: Let me try that. I see StartTicketPurchase is also called from RPC, where we do not necessarily need to pass in the chain client.

File: loader/loader.go (hunk 32 above)
Comment (tuxcanfly) [inline loader/loader.go]: Let me check that.

Comment (tuxcanfly) [inline loader/loader.go]: As discussed, added a nil check on chainClient here.

File: rpc/documentation/api.md (hunk 14 above)
Comment (tuxcanfly) [inline rpc/documentation/api.md]: Docs are up-to-date now.

**Comment by tuxcanfly** [comment]
Squashed. Should be good to go.
